lec07-pop-models.Rmd

```diff
@@ -1,4 +1,657 @@
+(either positive or negative) but the true solution doesn't. In the logistic equation example above,
+this numerical instability can happen if $N$ falls below $0$: the population will keep getting
+more and more negative after that. 
+
+## Using `deSolve` in R to solve differential equations numerically
+
+R has a package called `deSolve` which contains several functions to numerically solve differential equations.
+Let's look at the documentation for `deSolve`:
+
+```{r}
+library(deSolve)
+```
+```{r, eval=FALSE}
+help(deSolve)
+```
+
+This is pretty hard to read, especially if your calculus was a long time ago. The important thing to know for
+our purposes is that the equations we're working with are **first-order ordinary differential equations**.
+*First-order* means that the derivatives are all first derivatives (as opposed to second derivatives, third
+derivatives, etc.). **Ordinary** means that the derivatives are full derivatives (as opposed to partial
+derivatives which are written $\partial N / \partial t$). 
+
+This means that the function we'll be using is called `ode`, which stands for 'ordinary differential equation'. 
+
+```{r, eval=FALSE}
+help(ode)
+# The possible methods are all different ways of numerically solving DEs.
+
+#stiff vs. non-stiff : no precise definition, but 'stiff' generally means it's possible to be numerically unstable if the step size is too big. 
+```
+
+The method used by `ode` to numerically solve differential equations is called `lsoda`, which automatically
+tries to use the best method depending on the equation. This is why people like to use functions like `ode`:
+the result is often more accurate than Euler's method and might also be faster to run. 
+
+```{r}
+# Example: deSolve with logistic eqn
+
+# now we need to define our function a bit differently to be in the format that `ode` uses
+logistic_fn <- function(t,state,parameters) { 
+  # Calculates dN/dt for the logistic equation
+  
+  # t: time point at which to evaluate derivative (doesn't actually change anything in this example)
+  # state: vector of variables (here it's just N)
+  # parameters: vector of parameters c(r, K)
+  
+  N <- state 
+  
+  r <- parameters[1] # the first element of the parameters vector is r
+  K <- parameters[2] # the second element of the parameters vector is K
+
+  #rate of change
+  dN <- r*N*(1-N/K)
+    
+  #return rate of change
+  list(c(dN))
+}
+
+parameters <- c(r = 0.5, K = 50)
+state <- c(N = 10)
+times <- seq(0, 50, by = 0.01)
+
+#help(ode) # look at the documentation to learn about the parameters
+result <- ode(y = state, times = times, func = logistic_fn, parms = parameters)
+```
+
+The output of `ode` is a vector that contains the times we requested and its calculated values of N. 
+
+Notice that by using `ode` we didn't have to write a loop explicitly ourselves, and we didn't have to
+choose a timestep. `ode` does under the hood a version of what we did explicitly before. 
```

Comment:
Change

> `ode` does under the hood a version of what we did explicitly before.

to

> Under the hood, `ode` performs a similar computation to what we did manually. The `ode` version of this calculation is faster than our implementation, so we save both time writing and executing the code.

Reply:
It's not necessarily faster to execute, but this isn't something I know a lot about.